finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
+    BE: Backend<B> + 'static,
+    SC: SelectChain<B> + 'static,
+{
     select_chain: SC,
     client: Arc<C>,
-    proposed_block: Arc<Mutex<AlephDataFor<B>>>,
+    data_to_propose: Arc<Mutex<AlephData<B>>>,
+    min_block_num: NumberFor<B>,
     max_block_num: NumberFor<B>,
-    mut exit: oneshot::Receiver<()>,
-) where
+    best_block_in_session: Option<BlockHashNum<B>>,
+    prev_chain_info: Option<ChainInfoForData<B>>,
+    _phantom: PhantomData<BE>,
+}
+
+impl<B, BE, SC, C> ChainTracker<B, BE, SC, C>
+where
     B: BlockT,
     C: crate::ClientForAleph<B, BE> + Send + Sync + 'static,
     BE: Backend<B> + 'static,
     SC: SelectChain<B> + 'static,
 {
-    // We would like proposed_block to contain the highest ancestor of the `best_block` (this is what
-    // `select_chain` provides us with) up to the maximal height of `max_block_num`. This task periodically
-    // queries `select_chain` for the `best_block` and updates proposed_block accordingly. One optimization that it
-    // uses is that once the block in `proposed_block` reaches the height of `max_block_num`, and the just queried
-    // `best_block` is a `descendant` of the previous query, then we don't need to update proposed_block, as it is
-    // already correct.
-    let mut prev_best_hash: B::Hash = proposed_block.lock().hash;
-    let mut prev_best_number: NumberFor<B> = proposed_block.lock().number;
-    loop {
-        let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
-        tokio::select! {
-            _ = delay => {
-                let new_best_header = select_chain
-                    .best_chain()
-                    .await
-                    .expect("No best chain");
-                if *new_best_header.number() <= max_block_num {
-                    *proposed_block.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
-                } else {
-                    // we check if prev_best_header is an ancestor of new_best_header:
-                    if proposed_block.lock().number < max_block_num {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                        *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                    } else {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), prev_best_number);
-                        if reduced_header.hash() != prev_best_hash {
-                            // the new best block is not a descendant of previous best
-                            let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                            *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                        }
-                    }
+    pub(crate) fn new(
+        select_chain: SC,
```

Comment:
Yes, that's a good question. If you start digging into substrate then there is a place where they say you should use `SelectChain` instead of what you are proposing. Why is that, I have no idea. It annoys me a lot actually :P

Reply:
As far as I understand, `best_block` from `HeaderBackend` corresponds to the currently highest one; `best_block` from `SelectChain` corresponds to semantically best one, i.e. given a set of leaves, it returns the best one, whatever 'best' mean;

If this is the case, then using `SelectChain` should be encouraged in places where extending chain is the goal, and `HeaderBackend` should take precedence when just checking chain state.

However, indeed Substrate says: "_Usages that require determining an accurate best block should use `SelectChain` instead of the client._". Therefore, I'm kinda confused :|

Reply:
Yes, that's exactly the comment I'm referring to. I would be happy to ditch this `select_chain` thing, but I'm afraid something will happen :P

Reply:
But also: if that's what we want to do (and generally I think we do), then I would be in favor or creating a separate task that consists of research + implementation.